react/more-react-concepts/managing-state-with-context-api/managing-state-with-context-api.md

```diff
@@ -0,0 +1,280 @@
+
+~~~jsx
+import { useContext } from 'react';
+// import for ShopContext
+// import for Link
+
+function Links() {
+  const { cartItems } = useContext(ShopContext); // We must pass the ShopContext object itself as an argument
+
+  return (
+    <ul>
+      {/* Other links */}
+      <li>
+        <Link to="Link to the cart">
+          <span>Cart</span>
+          <div class="cart-icon">{cartItems.length}</div>
+        </Link>
+      </li>
+    </ul>
+  );
+}
+
+export default function Header() {
+  return (
+    <header>
+      {/* Other header elements */}
+      <nav>
+        <Links />
+      </nav>
+    </header>
+  );
+}
+~~~
+
+We've completely removed the prop drill problem, and we can conveniently get the `cartItems` directly in the `Links` component itself as we already know that no matter how deeply nested the component is, we can still get the data as long as _it's nested inside the Provider_.
+
+Let's also change our `ProductDetail` component to do the same:
+
+~~~jsx
+import { useContext } from 'react';
+// import for ShopContext
+
+export default function ProductDetail() {
+  const { products, addToCart } = useContext(ShopContext);
+  const product = products.find(/* Logic to find the specific product */);
+
+  return (
+    <div>
+      {/* Image of the product */}
+      <div>
+        {/* elements that align with the design */}
+        <button type="button" onClick={() => addToCart(product)}>
+          Add to Cart
+        </button>
+      </div>
+    </div>
+  );
+}
+~~~
+
+<a name="context-api-benefits"></a>
+In the `Header` component, we used `useContext()` to access `cartItems` from the `ShopContext`. Similarly, in the `ProductDetail` component, we can use the `products` and `addToCart` function.
+
+With this implementation, we no longer need to pass down props from the `App` component to the `Header` and `ProductDetail` components. The `Header` and `ProductDetail` components can simply access the `cartItems` state and `addToCart` function from the `ShopContext`.
+
+Overall, the implementation of the Context API has allowed for a more efficient, cleaner, and streamlined way of passing down data across multiple components. By using the `createContext()` function and the `useContext()` hook, we can easily pass down state and functions to child components without the need for prop drilling. It's also more centralized since we're keeping our data in a single location. All of these make our code easier to reason about.
+
+### Drawbacks of using Context API
+
+Although the Context API can be a powerful tool for managing state in larger React applications, it also has some drawbacks that you should be aware of:
+
+1. It can lead to performance issues: When you update the state in a context, it can cause all components that are consuming that context to re-render, even if the state that they are using hasn't changed. This can lead to performance issues, especially if you have a lot of components that are using the same context.
```

Comment:
Wording suggestion
```suggestion
1. It can lead to performance issues: When you update the state in a context, it can cause all components that are consuming that context to re-render, even if the state that they are using hasn't changed. This can lead to performance issues, especially if you have a lot of components that are consuming the same context.
```

Reply:
Changed the 2nd one as well with the same word